src/main/java/com/google/devtools/build/lib/rules/cpp/CcBinary.java

```diff
@@ -1046,13 +1048,29 @@ private static Packager createIntermediateDwpPackagers(
   private static NestedSet<Artifact> createDynamicLibrariesCopyActions(
       RuleContext ruleContext, Iterable<Artifact> dynamicLibrariesForRuntime) {
     NestedSetBuilder<Artifact> result = NestedSetBuilder.stableOrder();
+    CppConfiguration cppConfiguration = ruleContext.getFragment(CppConfiguration.class);
     for (Artifact target : dynamicLibrariesForRuntime) {
       if (!ruleContext
           .getLabel()
           .getPackageIdentifier()
-          .equals(target.getOwner().getPackageIdentifier())) {
+          .equals(target.getOwner().getPackageIdentifier())
+          || cppConfiguration.isRenameDLL()) {
```

Comment:
We have to check `isRenameDLL()` here since we wrote `name_hash.dll` instead of `name.dll` to the DEF file in the first place. It will rename the dll to `name_hash.dll` even if the dll is in the same directory. It causes a problem when there are many rules depend on the same dll but we could check if `name_hash.dll` is existed before rename.
```
cc_binary(
  name = "a",
  deps = [":c"],
)

cc_binary(
  name = "b",
  deps = [":c"],
)

cc_library(
  name = "c",
)
```
If we want to use the truly benefit of shared linked library (i.e one dll for many binary/library), we may want to create a folder to store dll and `set PATH=c:\bazel\MyDLLFolder;%PATH%`. `set` applies to the current command prompt only so we will be fine

Reply:
I think it's better to do the renaming in CcLibrary when we generate the DLL than in CcBinary when we copy the DLLs. Because

- `dynamicLibrariesForRuntime` may contain not only DLLs generated by CcLibrary but also DLL imported by ` cc_import` rule. If we rename the DLL, it won't work.
- Ww put the renamed DLL name in DEF file in CcLibrary, which will generate a LIB file (interface library) that is not consistent with the DLL generated by the same CcLibrary rule. I think that's not a good thing.

In CcLibrary, you can probably add a `linkedArtifactNameSuffix` pretty easily here:
https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibrary.java;l=629?q=CcLibrary.java&ss=bazel%2Fbazel

To make sure `--rename_dll` flag only affects Windows build, you can always check `featureConfiguration.isEnabled(CppRuleClasses.TARGETS_WINDOWS)`.

Reply:
Thank you for your reviews. And how about the idea creating a folder to store dll and `set PATH=c:\bazel\MyDLLFolder;%PATH%`. If it is good, I will implement it in this PR too and we could change the flag to something like `--better_dll_support`